src/nvim/api/autocmd.c

```diff
@@ -0,0 +1,669 @@
+                                  au_group,
+                                  is_once,
+                                  is_nested,
+                                  desc,
+                                  aucmd);
+      });
+
+      if (retval == FAIL) {
+        api_set_error(err, kErrorTypeException, "Failed to set autocmd");
+        goto cleanup;
+      }
+    }
+  });
+
+
+cleanup:
+  aucmd_exec_free(&aucmd);
+  api_free_array(event_array);
+  api_free_array(patterns);
+
+  return autocmd_id;
+}
+
+/// Delete an autocmd by ID. Autocmds only return IDs when created
+/// via the API.
+///
+/// @param id Integer The ID returned by nvim_create_autocmd
+void nvim_del_autocmd(Integer id)
+  FUNC_API_SINCE(9)
+{
+  autocmd_delete_id(id);
+}
+
+/// Create or get an augroup.
+///
+/// To get an existing augroup ID, do:
+/// <pre>
+///     local id = vim.api.nvim_create_augroup({ name = name, clear = false });
+/// </pre>
+///
+/// @param opts Parameters
+///                 - name (string): The name of the augroup
+///                 - clear (bool): Whether to clear existing commands or not.
+//                                  Defaults to true.
+///                     See |autocmd-groups|
+Integer nvim_create_augroup(uint64_t channel_id, Dict(create_augroup) *opts, Error *err)
+  FUNC_API_SINCE(9)
+{
+  bool clear_autocmds = api_object_to_bool(opts->clear, "clear", true, err);
+
+  if (opts->name.type != kObjectTypeString) {
+    api_set_error(err, kErrorTypeValidation, "'name' is required and must be a string");
+    return -1;
+  }
+  char *name = opts->name.data.string.data;
+
+  int augroup = -1;
+  WITH_SCRIPT_CONTEXT(channel_id, {
+    augroup = augroup_add(name);
+    if (augroup == AUGROUP_ERROR) {
+      api_set_error(err, kErrorTypeException, "Failed to set augroup");
+      return -1;
+    }
+
+    if (clear_autocmds) {
+      FOR_ALL_AUEVENTS(event) {
+        aupat_del_for_event_and_group(event, augroup);
+      }
+    }
+  });
+
+  return augroup;
+}
+
+/// NOTE: behavior differs from augroup-delete.
+/// When deleting an augroup, autocmds contained by this augroup will also be deleted and cleared.
+/// This augroup will no longer exist
+void nvim_del_augroup_by_id(Integer id)
+  FUNC_API_SINCE(9)
+{
+  char *name = augroup_name((int)id);
+  augroup_del(name, false);
+}
+
+/// NOTE: behavior differs from augroup-delete.
+/// When deleting an augroup, autocmds contained by this augroup will also be deleted and cleared.
+/// This augroup will no longer exist
+void nvim_del_augroup_by_name(String name)
+  FUNC_API_SINCE(9)
+{
+  augroup_del(name.data, false);
+}
+
+/// -- @param {string} group - autocmd group name
+/// -- @param {number} buffer - buffer number
+/// -- @param {string | table} event - event or events to match against
+/// -- @param {string | table} pattern - optional, defaults to "*".
+/// vim.api.nvim_do_autcmd({ group, buffer, pattern, event, modeline })
+void nvim_do_autocmd(Dict(do_autocmd) *opts, Error *err)
```

Comment:
`event` is mandatory for this function anything at all right? so could be positional.